server/src/main/java/io/druid/client/HttpServerInventoryView.java

```diff
@@ -0,0 +1,719 @@
+                  finalPredicate = Predicates.or(
+                      defaultFilter,
+                      Predicates.or(segmentPredicates.values())
+                  );
+                }
+              }
+            }
+          }
+      );
+    }
+  }
+
+  private void runServerCallbacks(final DruidServer server)
+  {
+    for (final Map.Entry<ServerCallback, Executor> entry : serverCallbacks.entrySet()) {
+      entry.getValue().execute(
+          new Runnable()
+          {
+            @Override
+            public void run()
+            {
+              if (CallbackAction.UNREGISTER == entry.getKey().serverRemoved(server)) {
+                serverCallbacks.remove(entry.getKey());
+              }
+            }
+          }
+      );
+    }
+  }
+
+  //best effort wait for first segment listing fetch from all servers and then call
+  //segmentViewInitialized on all registered segment callbacks.
+  private void serverInventoryInitialized()
+  {
+    for (DruidServerHolder server : servers.values()) {
+      server.awaitInitialization();
+    }
+
+    log.info("Calling SegmentCallback.segmentViewInitialized() for all callbacks.");
+
+    runSegmentCallbacks(
+        new Function<SegmentCallback, CallbackAction>()
+        {
+          @Override
+          public CallbackAction apply(SegmentCallback input)
+          {
+            return input.segmentViewInitialized();
+          }
+        }
+    );
+  }
+
+  private DruidServer serverAddedOrUpdated(DruidServer server)
+  {
+    DruidServerHolder curr;
+    DruidServerHolder newHolder;
+    synchronized (servers) {
+      curr = servers.get(server.getName());
+      newHolder = curr == null ? new DruidServerHolder(server) : curr.updatedHolder(server);
+      servers.put(server.getName(), newHolder);
+    }
+
+    //fetch list of segments
+    if (curr == null) {
+      try {
+        Future<?> future = newHolder.updateSegmentsListAsync();
+        future.get(config.getServerTimeout(), TimeUnit.MILLISECONDS);
+      } catch (Throwable th) {
+        log.error(th, "Failed to fetch segment list from %s , will be retried", server.getName());
+      }
+    }
+
+    return newHolder.druidServer;
+  }
+
+  private void serverRemoved(DruidServer server)
+  {
+    synchronized (servers) {
+      servers.remove(server.getName());
+    }
+  }
+
+  public DruidServer serverUpdated(DruidServer oldServer, DruidServer newServer)
+  {
+    return serverAddedOrUpdated(newServer);
+  }
+
+  @Override
+  public boolean isStarted()
+  {
+    return started.get();
+  }
+
+  @Override
+  public boolean isSegmentLoadedByServer(String serverKey, DataSegment segment)
+  {
+    synchronized (servers) {
+      DruidServerHolder holder = servers.get(serverKey);
+      return holder.druidServer.getSegment(segment.getIdentifier()) != null;
+    }
+  }
+
+  private class DruidServerHolder
+  {
+    private final Object lock = new Object();
+
+    @GuardedBy("lock")
```

Comment:
added comment in code, this is to ensure consistent segment state stored in DruidServer and counter managed.